### PR TITLE
Pull Request for Issue 1566: Continuous Energy Motion

### PR DESCRIPTION
**Pull request by iainworkman**
- Added ability to do continuous trajectory movements of energy to the pseudomotor
- Added PVs for trajectory movement to the db file
- Added the trajectory movement PVs to the coordinator, in order to pass along the move to the pseudomotor
- Added the new trajectory movement PVs to Acquaman, so they can be written to.
- Added a new view for controlling the trajectory motion, and added it to the SGMEnergyView.

#1566

### CODE REVIEW

File: source/beamline/SGM/SGMEnergyCoordinator.cpp [hunk 1]
```diff
@@ -556,7 +559,8 @@ void SGMEnergyCoordinator::onEnergySetpointPVChanged(double value)
 	if(energySetpointInitialized_)  {
 
 		qDebug() << QString("Forwarding energy setpoint value of %1 to energy control").arg(value);
-		energyControlCoordinator_->move(value);
+		AMControl* controlVersion = energyControlCoordinator_;
```
Comment (dretrex) [inline source/beamline/SGM/SGMEnergyCoordinator.cpp]: Is there a reason this isn't just `energyControlCoordinator_->move(value);`?

File: source/beamline/SGM/SGMEnergyCoordinator.cpp [hunk 2]
```diff
@@ -576,6 +580,27 @@ void SGMEnergyCoordinator::onEnergyStopPVChanged(double)
 	}
 }
 
+void SGMEnergyCoordinator::onEnergyTrajectoryStartPVChanged(double value)
+{
+	qDebug() << "\tTrajectory start value of" << value << "received";
```
Comment (dretrex) [inline source/beamline/SGM/SGMEnergyCoordinator.cpp]: Are these the debugs for terminal information or just testing ones?

File: source/beamline/SGM/energy/SGMEnergyControl.cpp [hunk 3]
```diff
@@ -162,65 +152,162 @@ AMControl *SGMEnergyControl::gratingTranslationControl() const
...
+
+	moveAction->start();
+
+	return AMControl::NoFailure;
 }
 
 void SGMEnergyControl::setUndulatorHarmonic(SGMUndulatorSupport::UndulatorHarmonic undulatorHarmonic)
 {
-    if(energyPositionController_) {
+	if(energyPositionController_) {
```
Comment (dretrex) [inline source/beamline/SGM/energy/SGMEnergyControl.cpp]: We really need to get these tabs sorted out.

File: source/beamline/SGM/SGMEnergyCoordinator.cpp (hunk 1 above)
Comment (iainworkman) [inline source/beamline/SGM/SGMEnergyCoordinator.cpp]: It is to stop the initial setting of the energy during start up from being actually prompting the energy control from making a full move.
I did this because there may be a case where someone moves to a given energy with undulator tracking off, then for some reason restarts the coordinator. If this check wasn't here when the energy is calculated from the grating on start up, and this value written to the energy setpoint (to ensure it's at the right value to being with) the undulator (which because tracking was off during the initial move to this position isn't at the right spot) would begin to move. 

In short I didn't want a situation to arise where the undulator would begin to move just because the application restarted.

File: source/beamline/SGM/SGMEnergyCoordinator.cpp (hunk 2 above)
Comment (iainworkman) [inline source/beamline/SGM/SGMEnergyCoordinator.cpp]: This one was just for testing. I will remove it.

File: source/beamline/SGM/energy/SGMEnergyControl.cpp (hunk 3 above)
Comment (iainworkman) [inline source/beamline/SGM/energy/SGMEnergyControl.cpp]: Yep. Sorry :/

File: source/beamline/SGM/energy/SGMEnergyTrajectory.cpp [hunk 4]
```diff
@@ -1,199 +1,218 @@
...
+		startEnergyPosition_->setAutoDetectUndulatorHarmonic(false);
+		endEnergyPosition_->setAutoDetectUndulatorHarmonic(false);
+
+		startEnergyPosition_->setUndulatorHarmonic(SGMUndulatorSupport::FirstHarmonic);
+		endEnergyPosition_->setUndulatorHarmonic(SGMUndulatorSupport::FirstHarmonic);
+	}
+
+	time_ = time;
+
+	gratingAngleVelocityProfile_ = AMTrapezoidVelocityProfile(startEnergyPosition_->gratingAngle() * gratingAngleStepsPerEncoderCount,
```
Comment (dretrex) [inline source/beamline/SGM/energy/SGMEnergyTrajectory.cpp]: The conversion here is due to the fact that the acceleration and velocity are in steps but all of the positions are in real units, right?

Comment (iainworkman) [inline source/beamline/SGM/energy/SGMEnergyTrajectory.cpp]: Yes, that is correct.

File: source/beamline/SGM/energy/SGMEnergyTrajectory.cpp [hunk 5]
```diff
@@ -1,199 +1,218 @@
...
 double SGMEnergyTrajectory::endEnergy() const
 {
-    return endEnergy_;
+	return endEnergy_;
+}
+
+double SGMEnergyTrajectory::energyVelocity() const
+{
+	if(time_ != 0) {
+		return qAbs(endEnergy_ - startEnergy_) / time_;
```
Comment (dretrex) [inline source/beamline/SGM/energy/SGMEnergyTrajectory.cpp]: It is okay that this can return negative numbers, right?

File: source/beamline/SGM/energy/SGMGratingAngleControl.cpp [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+	                                                                    "Moving Grating Angle"),
+	                                              AMListAction3::Sequential);
+
+
+	if(isClosedLoop()) {
+
+		moveAction->addSubAction(AMActionSupport::buildControlMoveAction(encoderControl_, setpoint));
+		moveAction->addSubAction(AMActionSupport::buildControlWaitAction(encoderControl_, setpoint, 20, AMControlWaitActionInfo::MatchWithinTolerance));
+	} else {
+		;
```
Comment (dretrex) [inline source/beamline/SGM/energy/SGMGratingAngleControl.cpp]: Rogue ;!  Get 'em!

File: source/beamline/SGM/energy/SGMUndulatorControl.h [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+#define DEFAULT_UNDULATOR_VELOCITY 5000
+#define DEFAULT_UNDULATOR_ACCELERATION  2500
+
+#include "beamline/AMPseudoMotorControl.h"
+
+class SGMUndulatorControl : public AMPseudoMotorControl
+{
+	Q_OBJECT
+public:
+	explicit SGMUndulatorControl(QObject *parent = 0);
```
Comment (dretrex) [inline source/beamline/SGM/energy/SGMUndulatorControl.h]: It looks like the `SGMUndulatorControl` and the `SGMGratingControl` are almost identical.  As a future cleanup, I'm pretty sure they could inherit from a single class that does all the work.

File: source/beamline/SGM/energy/SGMEnergyTrajectory.cpp (hunk 5 above)
Comment (iainworkman) [inline source/beamline/SGM/energy/SGMEnergyTrajectory.cpp]: Will double check that. I think it should be an absolute value.

File: source/ui/SGM/SGMEnergyTrajectoryView.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+	Q_OBJECT
+public:
+	explicit SGMEnergyTrajectoryView(SGMEnergyControlSet* controlSet,
+	                                 QWidget *parent = 0);
+
+signals:
+
+public slots:
+protected slots:
+	void onStartPushButtonClicked();
```
Comment (dretrex) [inline source/ui/SGM/SGMEnergyTrajectoryView.h]: A couple header comments would be good.

File: source/util/AMTrapezoidVelocityProfile.cpp [hunk 9]
```diff
@@ -1,49 +1,62 @@
...
-    return targetVelocity_;
+	return time_;
 }
 
-double AMTrapezoidVelocityProfile::returnAcceleration() const
+double AMTrapezoidVelocityProfile::targetVelocity() const
 {
-    return returnAcceleration_;
+	double halfAccelByTime = (acceleration_*time_) / 2;
+	double toRoot = pow(halfAccelByTime, 2) - acceleration_ * qAbs(endPosition_ - startPosition_);
```
Comment (dretrex) [inline source/util/AMTrapezoidVelocityProfile.cpp]: Why `qAbs()`? I'm trying to remember a function that would require the qAbs and can't put my finger on it.

File: source/util/AMTrapezoidVelocityProfile.cpp [hunk 10]
```diff
@@ -1,49 +1,62 @@
...
+
+	if(toRoot < 0) {
+		return halfAccelByTime + sqrt(qAbs(toRoot));
```
Comment (dretrex) [inline source/util/AMTrapezoidVelocityProfile.cpp]: Same here.  Something seems a bit fishy if you need to qAbs this to me.

File: source/beamline/SGM/energy/SGMEnergyTrajectory.cpp (hunk 5 above)
Comment (iainworkman) [inline source/beamline/SGM/energy/SGMEnergyTrajectory.cpp]: This is fine, I believe. The validator should pick up cases where the time < 0.

File: source/beamline/SGM/SGMEnergyCoordinator.cpp (hunk 1 above)
Comment (dretrex) [inline source/beamline/SGM/SGMEnergyCoordinator.cpp]: That makes sense with with the setpoint initialization, but you are also setting the control here for no reason.

**Comment by dretrex** [comment]
Overall, this looks good.  But I am going to reiterate that we need to get those tabs sorted out.  That was probably >4000 lines of impact in tabs alone and very time consuming.
Time Estimate: 35 minutes

File: source/util/AMTrapezoidVelocityProfile.cpp (hunk 10 above)
Comment (iainworkman) [inline source/util/AMTrapezoidVelocityProfile.cpp]: This was arrived on with help from David. Essentially this bottom case is when the acceleration is not sufficient to reach the target velocity within the delta T. In that case this simply needs to ensure its a value above that peak, which this does.

Comment (dretrex) [inline source/util/AMTrapezoidVelocityProfile.cpp]: Okay. Sounds good.
